# Assignment 1: Python Fundamentals for Data Streaming

## AITU Campus Shuttle and Mobility Data

**Student:** Aliya Kuanysh  
**Program:** Big Data Analytics  
**Group:** BDA-2404

This notebook demonstrates data validation, generator-based streaming, JSON conversion, FastAPI integration and basic analysis of synthetic campus shuttle events.

In [1]:
from shuttle_data import RAW_EVENTS, validate_event, event_generator

In [2]:
for event in RAW_EVENTS:
    print(event)

{'Timestamp': '08:00', 'Vehicle_ID': 'AITU-Campus', 'Location': 'Residence B01', 'Passengers': 18, 'Speed_kmh': 31, 'Status': 'ON_ROUTE'}
{'Timestamp': '08:01', 'Vehicle_ID': 'AITU-Campus', 'Location': 'Residence B02', 'Passengers': 22, 'Speed_kmh': 28, 'Status': 'ON_ROUTE'}
{'Timestamp': '08:02', 'Vehicle_ID': 'AITU-Campus', 'Location': 'Residence B01', 'Passengers': 21, 'Speed_kmh': 29, 'Status': 'ON_ROUTE'}
{'Timestamp': '08:03', 'Vehicle_ID': 'AITU-Campus', 'Location': 'Residence B02', 'Passengers': 25, 'Speed_kmh': 27, 'Status': 'ON_ROUTE'}
{'Timestamp': '08:04', 'Vehicle_ID': 'AITU-Campus', 'Location': 'Residence B01', 'Passengers': 24, 'Speed_kmh': 0, 'Status': 'STOPPED'}
{'Timestamp': '08:05', 'Vehicle_ID': 'AITU-Campus', 'Location': 'Residence B02', 'Passengers': 26, 'Speed_kmh': 30, 'Status': 'ON_ROUTE'}
{'Timestamp': '08:06', 'Vehicle_ID': 'AITU-Campus', 'Location': 'Residence B01', 'Passengers': 23, 'Speed_kmh': 32, 'Status': 'ON_ROUTE'}
{'Timestamp': '08:07', 'Vehicle_ID':

In [3]:
validated_events = []

for event in RAW_EVENTS:
    validated_event = validate_event(event)
    validated_events.append(validated_event)

print(f"Validated events: {len(validated_events)}")

Validated events: 10


## Generator-Based Streaming

The `event_generator()` function uses `yield` to provide events one at a time. This approach is suitable for continuous data streams because the whole stream does not need to be stored in memory.

In [4]:
stream = event_generator()

for i in range(3):
    print(next(stream))

{'Timestamp': datetime.time(8, 0), 'Vehicle_ID': 'AITU-Campus', 'Location': 'Residence B01', 'Passengers': 18, 'Speed_kmh': 31, 'Status': 'ON_ROUTE'}
{'Timestamp': datetime.time(8, 1), 'Vehicle_ID': 'AITU-Campus', 'Location': 'Residence B02', 'Passengers': 22, 'Speed_kmh': 28, 'Status': 'ON_ROUTE'}
{'Timestamp': datetime.time(8, 2), 'Vehicle_ID': 'AITU-Campus', 'Location': 'Residence B01', 'Passengers': 21, 'Speed_kmh': 29, 'Status': 'ON_ROUTE'}


In [5]:
from shuttle_analysis import analyze_events

results = analyze_events()

results

{'average_passengers': 24.4,
 'maximum_passengers': 30,
 'stopped_count': 1,
 'busiest_minute': '08:09',
 'busiest_vehicle': 'AITU-Campus',
 'busiest_location': 'Residence B02',
 'busiest_passengers': 30}

In [6]:
from shuttle_api import get_occupancy_category

for passengers in [8, 15, 25, 35]:
    print(
        passengers,
        "passengers ->",
        get_occupancy_category(passengers)
    )

8 passengers -> LOW
15 passengers -> MEDIUM
25 passengers -> HIGH
35 passengers -> OVER_CAPACITY


## Conclusion

The project successfully demonstrates the basic processing of a streaming dataset.

The events were validated, generated one at a time, converted to JSON, analysed and exposed through a FastAPI endpoint.

The average passenger count was 24.4, the maximum was 30, and one event had STOPPED status. The busiest event occurred at 08:09 at Residence B02 with 30 passengers.

The main implementation decision was to use generators for event streaming because they allow events to be processed one at a time instead of loading a potentially unlimited stream into memory.